### PR TITLE
ITKDev - Patron registration

### PR DESCRIPTION
**Pull request by cableman**
#### Link to issue

https://platform.dandigbib.org/issues/5369

#### Description

Allow patrons to performance self registration.

#### Screenshot of the result

N/A

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

### CODE REVIEW

File: web/modules/custom/dpl_patron_reg/src/DplPatronRegSettings.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Class that handles patron registration settings.
+ */
+class DplPatronRegSettings extends DplReactConfigBase {
+
+  /**
+   * Gets the configuration key for the instant patron registration settings.
+   */
+  public function getConfigKey(): string {
+    return 'dpl_instant_loan.settings';
```
Comment (spaceo) [inline web/modules/custom/dpl_patron_reg/src/DplPatronRegSettings.php:16]: This should be a config key reserved for dpl patron settings

Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/DplPatronRegSettings.php:16]: My bad

Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/DplPatronRegSettings.php]: Fixed

**Review by spaceo — CHANGES_REQUESTED** [review]
Nice work! 👍  I have some comments for things I think should be changed.

File: web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public function getFormId(): string {
+    return 'dpl_patron_reg_settings_form';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state): array {
+    $config = $this->configService->getConfig();
```
[the remark this answers is not in the document]
Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php]: Fixed

File: web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    // the URL.
+    $url = UrlHelper::parse($response->getTargetUrl());
+    $url['query']['idp'] = 'nemlogin';
+    $url = Url::fromUri($url['path'], ['query' => $url['query']]);
+    $url->setAbsolute();
+    $url = $url->toString();
+
+    /** @var string $url */
+    return new TrustedRedirectResponse($url);
+  }
```
[the remark this answers is not in the document]
Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php:143]: That is partially correct as we here need to enforce nemlogin/mitid and therefor can not use the once you link to. So this is somewhat at special case.

[the remark this answers is not in the document]
Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php:143]: Also the parameter used to enforce mitid is out side the oAuth standard spec. and therefor require manual building the redirect URL.

File: web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  public function userRegistrationReactAppLoad(): array {
+    /** @var \Drupal\dpl_patron_reg\Plugin\Block\PatronRegistrationBlock $plugin_block */
+    $plugin_block = $this->blockManager->createInstance('dpl_patron_reg_block', []);
+
+    // Some blocks might implement access check.
+    $access_result = $plugin_block->access($this->currentUser());
+    if (is_object($access_result) && $access_result->isForbidden() || is_bool($access_result) && !$access_result) {
+      throw new AccessDeniedHttpException();
+    }
```
[the remark this answers is not in the document]
Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php:162]: Added todo

File: web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (spaceo) [reply web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php:143]: Ok. I can see the difference. 👍

File: web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      '#description' => $this->t('Redirect to this on user successful created'),
+      '#default_value' => $config->get('redirect_on_user_created_url') ?? '',
+    ];
+
+    $form['information'] = [
+      '#type' => 'text_format',
+      '#title' => $this->t('Information page'),
+      '#default_value' => $config->get('information')['value'] ?? '',
+      '#format' => $config->get('information')['format'] ?? 'plain_text',
+    ];
```
Comment (spaceo) [inline web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php:83]: How about validation of the fields?
Nothing here is required?
And is there a dependency between the data?
If yes it makes sense to use `#states` in the form fields to control what is depending on each other.
Example:
https://github.com/danskernesdigitalebibliotek/dpl-cms/blob/0c6d60916b6425243ef1c9b0ce3871b2eab9c4ee/web/modules/custom/dpl_instant_loan/src/Form/DplInstantLoanSettingsForm.php#L67-L109

File: web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (spaceo) [reply web/modules/custom/dpl_patron_reg/src/Controller/DplPatronRegController.php:162]: 👍

**Review by spaceo — COMMENTED** [review]
Look good.
One minor question...

**Review by spaceo — APPROVED** [review]
Approved but please look at my comment

File: web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php (hunk 5 above)
Comment (cableman) [reply web/modules/custom/dpl_patron_reg/src/Form/PatronRegSettingsForm.php:83]: The age field is limit to positive int and the URL is an URL filed and text is text. So they will be validated by default.

**Pull request merged**